Parallel Dataproviders & retries causes test result count to be skewed

### TestNG Version
> Note: only the latest version is supported

7.8.0


### Expected behavior

A test method that can be retried and which is powered by a data provider which can run in parallel is executed, then the count of the test results (total and individual count for passed, failed and skipped) should match.

### Actual behavior

The count is skewed

### Is the issue reproducible on runner?

- [X] Shell
- [X] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [X] IntelliJ
- [ ] NetBeans


### Test case sample
> Please, share the test case (as small as possible) which shows the issue

<details>

<summary>Sample testcase</summary>

```java
import static org.assertj.core.api.Assertions.assertThat;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;
import org.testng.IAnnotationTransformer;
import org.testng.IRetryAnalyzer;
import org.testng.ITestContext;
import org.testng.ITestListener;
import org.testng.ITestResult;
import org.testng.annotations.DataProvider;
import org.testng.annotations.IDataProviderAnnotation;
import org.testng.annotations.Test;

public class MyTestCase {

  @Test(dataProvider = "testData", retryAnalyzer = RetryAnalyzer.class)
  public void API_dummyTest(int number, String text) {
    System.out.println(
        "[" + Thread.currentThread().getId() + "] This is for " + text + " " + number);
    assertThat(number % 2).isEqualTo(0);
  }


  @DataProvider(name = "testData")
  public Object[][] getData() {
    return new Object[][]{
        {1, "Test One"},
        {2, "Test Two"},
        {3, "Test Three"}
    };
  }

  public static class ToggleDataProvider implements IAnnotationTransformer {

    private final boolean isParallel;

    public ToggleDataProvider(boolean isParallel) {
      this.isParallel = isParallel;
    }

    @Override
    public void transform(IDataProviderAnnotation annotation, Method method) {
      if (isParallel) {
        annotation.setParallel(true);
      }
    }
  }

  public static class RetryAnalyzer implements IRetryAnalyzer {

    private final AtomicInteger counter = new AtomicInteger(0);

    @Override
    public boolean retry(ITestResult iTestResult) {
      return counter.getAndIncrement() < 3;
    }
  }

  public static class CoreListener implements ITestListener {

    private ITestContext context;


    @Override
    public void onTestFailure(ITestResult result) {
      printer(result, "FAILED");
    }

    @Override
    public void onTestSkipped(ITestResult result) {
      printer(result, "SKIPPED");
    }

    @Override
    public void onTestSuccess(ITestResult result) {
      printer(result, "passed");
    }

    private static void printer(ITestResult result, String prefix) {
      System.err.println(prefix + " for values " + Arrays.toString(result.getParameters()));
    }

    @Override
    public void onFinish(ITestContext context) {
      this.context = context;
      display("PASSED", context.getPassedTests().getAllResults());
      display("FAILED", context.getFailedTests().getAllResults());
      display("SKIPPED", context.getSkippedTests().getAllResults());
    }

    public int totalTests() {
      return context.getPassedTests().size() + context.getFailedTests().size() +
          context.getSkippedTests().size();
    }

    public Set<ITestResult> getPassedTests() {
      return context.getPassedTests().getAllResults();
    }

    public Set<ITestResult> getFailedTests() {
      return context.getFailedTests().getAllResults();
    }

    public Set<ITestResult> getSkippedTests() {
      return context.getSkippedTests().getAllResults();
    }
  }

  private static void display(String prefix, Set<ITestResult> results) {
    System.err.println(prefix + " combos");
    results
        .stream()
        .peek(each -> System.err.println(
            "Retried ? " + each.wasRetried() + " test data " + Arrays.toString(
                each.getParameters())))
        .map(each -> Arrays.toString(each.getParameters()))
        .forEach(System.err::println);
  }
}
```

</details>

<details>

<summary>Sample Test runner to reproduce the bug</summary>

```java
import static org.assertj.core.api.Assertions.assertThat;

import com.rationaleemotions.dp.MyTestCase.CoreListener;
import com.rationaleemotions.dp.MyTestCase.ToggleDataProvider;
import java.util.Arrays;
import org.assertj.core.api.SoftAssertions;
import org.testng.ITestResult;
import org.testng.TestNG;
import org.testng.annotations.Test;

public class TestRunner {

  @Test
  public void runTestWithDataProviderInParallel() {
    runTest(true);
  }

  @Test
  public void runTestWithDataProviderInSequential() {
    runTest(false);
  }

  private static void runTest(boolean isParallel) {
    TestNG testng = new TestNG();
    testng.setTestClasses(new Class[]{MyTestCase.class});
    CoreListener listener = new CoreListener();
    ToggleDataProvider transformer = new ToggleDataProvider(isParallel);
    testng.addListener(listener);
    testng.addListener(transformer);
    testng.setVerbose(2);
    testng.run();
    SoftAssertions softly = new SoftAssertions();

    softly.assertThat(listener.totalTests())
        .withFailMessage("We should have had 9 test results in total.")
        .isEqualTo(9);

    //Assert passed tests
    softly.assertThat(listener.getPassedTests())
        .withFailMessage("We should have had ONLY 1 passed test.")
        .hasSize(1);
    listener.getPassedTests()
        .forEach(each -> softly.assertThat(each.wasRetried())
            .withFailMessage("Passed test " + stringify(each) + " should NOT have been retried")
            .isFalse());

    //Assert failed tests
    assertThat(listener.getFailedTests())
        .withFailMessage("We should have had 2 failed tests.")
        .hasSize(2);
    listener.getFailedTests()
        .forEach(each -> softly.assertThat(each.wasRetried())
            .withFailMessage("Failed test " + stringify(each) + " should NOT have been retried")
            .isFalse());

    //Assert skipped tests
    assertThat(listener.getSkippedTests())
        .withFailMessage("We should have had 6 skipped tests due to retries.")
        .hasSize(6);
    listener.getSkippedTests()
        .forEach(each -> softly.assertThat(each.wasRetried())
            .withFailMessage("Skipped test " + stringify(each) + " should have been retried")
            .isTrue());
    softly.assertAll();
  }

  private static String stringify(ITestResult itr) {
    return "(" + Arrays.toString(itr.getParameters()) + ")";
  }
}

```

</details>

<details>

<summary>console output</summary>

```bash
[15] This is for Test One 1
[17] This is for Test Three 3
[16] This is for Test Two 2
passed for values [2, Test Two]
SKIPPED for values [1, Test One]
SKIPPED for values [3, Test Three]
[17] This is for Test Three 3
[15] This is for Test One 1
SKIPPED for values [3, Test Three]
FAILED for values [1, Test One]
[17] This is for Test Three 3
FAILED for values [3, Test Three]
PASSED combos
Retried ? false test data [2, Test Two]
[2, Test Two]
FAILED combos
Retried ? false test data [1, Test One]
[1, Test One]
Retried ? false test data [3, Test Three]
[3, Test Three]
SKIPPED combos
Retried ? true test data [3, Test Three]
[3, Test Three]
Retried ? true test data [3, Test Three]
[3, Test Three]
Retried ? true test data [1, Test One]
[1, Test One]
PASSED: com.rationaleemotions.dp.MyTestCase.API_dummyTest(2, "Test Two")
FAILED: com.rationaleemotions.dp.MyTestCase.API_dummyTest(1, "Test One")
java.lang.AssertionError: 
expected: 0
 but was: 1
	at com.rationaleemotions.dp.MyTestCase.API_dummyTest(MyTestCase.java:24)
	at java.base/java.lang.Thread.run(Thread.java:829)

FAILED: com.rationaleemotions.dp.MyTestCase.API_dummyTest(3, "Test Three")
java.lang.AssertionError: 
expected: 0
 but was: 1
	at com.rationaleemotions.dp.MyTestCase.API_dummyTest(MyTestCase.java:24)
	at java.base/java.lang.Thread.run(Thread.java:829)

RETRIED: com.rationaleemotions.dp.MyTestCase.API_dummyTest(3, "Test Three")
java.lang.AssertionError: 
expected: 0
 but was: 1
	at com.rationaleemotions.dp.MyTestCase.API_dummyTest(MyTestCase.java:24)
	at java.base/java.lang.Thread.run(Thread.java:829)

RETRIED: com.rationaleemotions.dp.MyTestCase.API_dummyTest(3, "Test Three")
java.lang.AssertionError: 
expected: 0
 but was: 1
	at com.rationaleemotions.dp.MyTestCase.API_dummyTest(MyTestCase.java:24)
	at java.base/java.lang.Thread.run(Thread.java:829)

RETRIED: com.rationaleemotions.dp.MyTestCase.API_dummyTest(1, "Test One")
java.lang.AssertionError: 
expected: 0
 but was: 1
	at com.rationaleemotions.dp.MyTestCase.API_dummyTest(MyTestCase.java:24)
	at java.base/java.lang.Thread.run(Thread.java:829)


===============================================
    Command line test
    Tests run: 1, Failures: 2, Skips: 0, Retries: 3
===============================================


===============================================
Command line suite
Total tests run: 6, Passes: 1, Failures: 2, Skips: 0, Retries: 3
===============================================


java.lang.AssertionError: We should have had 6 skipped tests due to retries.

	at com.rationaleemotions.dp.TestRunner.runTest(TestRunner.java:61)
	at com.rationaleemotions.dp.TestRunner.runTestWithDataProviderInParallel(TestRunner.java:17)

[1] This is for Test One 1
SKIPPED for values [1, Test One]
[1] This is for Test One 1
SKIPPED for values [1, Test One]
[1] This is for Test One 1
SKIPPED for values [1, Test One]
[1] This is for Test One 1
FAILED for values [1, Test One]
[1] This is for Test Two 2
passed for values [2, Test Two]
[1] This is for Test Three 3
SKIPPED for values [3, Test Three]
[1] This is for Test Three 3
SKIPPED for values [3, Test Three]
[1] This is for Test Three 3
SKIPPED for values [3, Test Three]
[1] This is for Test Three 3
FAILED for values [3, Test Three]
PASSED combos
Retried ? false test data [2, Test Two]
[2, Test Two]
FAILED combos
Retried ? false test data [1, Test One]
[1, Test One]
Retried ? false test data [3, Test Three]
[3, Test Three]
SKIPPED combos
Retried ? true test data [3, Test Three]
[3, Test Three]
Retried ? true test data [3, Test Three]
[3, Test Three]
Retried ? true test data [3, Test Three]
[3, Test Three]
Retried ? true test data [1, Test One]
[1, Test One]
Retried ? true test data [1, Test One]
[1, Test One]
Retried ? true test data [1, Test One]
[1, Test One]
PASSED: com.rationaleemotions.dp.MyTestCase.API_dummyTest(2, "Test Two")
FAILED: com.rationaleemotions.dp.MyTestCase.API_dummyTest(1, "Test One")
java.lang.AssertionError: 
expected: 0
 but was: 1
	at com.rationaleemotions.dp.MyTestCase.API_dummyTest(MyTestCase.java:24)
	at com.rationaleemotions.dp.TestRunner.runTest(TestRunner.java:33)
	at com.rationaleemotions.dp.TestRunner.runTestWithDataProviderInSequential(TestRunner.java:22)

FAILED: com.rationaleemotions.dp.MyTestCase.API_dummyTest(3, "Test Three")
java.lang.AssertionError: 
expected: 0
 but was: 1
	at com.rationaleemotions.dp.MyTestCase.API_dummyTest(MyTestCase.java:24)
	at com.rationaleemotions.dp.TestRunner.runTest(TestRunner.java:33)
	at com.rationaleemotions.dp.TestRunner.runTestWithDataProviderInSequential(TestRunner.java:22)

RETRIED: com.rationaleemotions.dp.MyTestCase.API_dummyTest(3, "Test Three")
java.lang.AssertionError: 
expected: 0
 but was: 1
	at com.rationaleemotions.dp.MyTestCase.API_dummyTest(MyTestCase.java:24)
	at com.rationaleemotions.dp.TestRunner.runTest(TestRunner.java:33)
	at com.rationaleemotions.dp.TestRunner.runTestWithDataProviderInSequential(TestRunner.java:22)

RETRIED: com.rationaleemotions.dp.MyTestCase.API_dummyTest(3, "Test Three")
java.lang.AssertionError: 
expected: 0
 but was: 1
	at com.rationaleemotions.dp.MyTestCase.API_dummyTest(MyTestCase.java:24)
	at com.rationaleemotions.dp.TestRunner.runTest(TestRunner.java:33)
	at com.rationaleemotions.dp.TestRunner.runTestWithDataProviderInSequential(TestRunner.java:22)

RETRIED: com.rationaleemotions.dp.MyTestCase.API_dummyTest(3, "Test Three")
java.lang.AssertionError: 
expected: 0
 but was: 1
	at com.rationaleemotions.dp.MyTestCase.API_dummyTest(MyTestCase.java:24)
	at com.rationaleemotions.dp.TestRunner.runTest(TestRunner.java:33)
	at com.rationaleemotions.dp.TestRunner.runTestWithDataProviderInSequential(TestRunner.java:22)

RETRIED: com.rationaleemotions.dp.MyTestCase.API_dummyTest(1, "Test One")
java.lang.AssertionError: 
expected: 0
 but was: 1
	at com.rationaleemotions.dp.MyTestCase.API_dummyTest(MyTestCase.java:24)
	at com.rationaleemotions.dp.TestRunner.runTest(TestRunner.java:33)
	at com.rationaleemotions.dp.TestRunner.runTestWithDataProviderInSequential(TestRunner.java:22)

RETRIED: com.rationaleemotions.dp.MyTestCase.API_dummyTest(1, "Test One")
java.lang.AssertionError: 
expected: 0
 but was: 1
	at com.rationaleemotions.dp.MyTestCase.API_dummyTest(MyTestCase.java:24)
	at com.rationaleemotions.dp.TestRunner.runTest(TestRunner.java:33)
	at com.rationaleemotions.dp.TestRunner.runTestWithDataProviderInSequential(TestRunner.java:22)

RETRIED: com.rationaleemotions.dp.MyTestCase.API_dummyTest(1, "Test One")
java.lang.AssertionError: 
expected: 0
 but was: 1
	at com.rationaleemotions.dp.MyTestCase.API_dummyTest(MyTestCase.java:24)
	at com.rationaleemotions.dp.TestRunner.runTest(TestRunner.java:33)
	at com.rationaleemotions.dp.TestRunner.runTestWithDataProviderInSequential(TestRunner.java:22)


===============================================
    Command line test
    Tests run: 1, Failures: 2, Skips: 0, Retries: 6
===============================================


===============================================
Command line suite
Total tests run: 9, Passes: 1, Failures: 2, Skips: 0, Retries: 6
===============================================


===============================================
Default Suite
Total tests run: 2, Passes: 1, Failures: 1, Skips: 0
===============================================


Process finished with exit code 0
```
</details>

### Contribution guidelines

Incase you plan to raise a pull request to fix this issue, please make sure you refer our [Contributing](.github/CONTRIBUTING.md) section for detailed set of steps.

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,5 @@
 Current
+Fixed: GITHUB-2934: Parallel Dataproviders & retries causes test result count to be skewed (Krishnan Mahadevan)
 Fixed: GITHUB-2925: Issue in ITestcontext.getAllTestMethods() with annotation @BeforeSuite (Krishnan Mahadevan)
 Fixed: GITHUB-2928: The constructor of TestRunner encountered NBC changes in 7.8.0 (Krishnan Mahadevan)
 Fixed: GITHUB-581: Parameters of nested test suites are overridden(Krishnan Mahadevan)

--- a/testng-core/src/main/java/org/testng/internal/BaseTestMethod.java
+++ b/testng-core/src/main/java/org/testng/internal/BaseTestMethod.java
@@ -6,6 +6,7 @@ import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.Callable;
@@ -817,7 +818,8 @@ public abstract class BaseTestMethod implements ITestNGMethod, IInvocationStatus
       this.m_retryAnalyzer = computeRetryAnalyzerInstanceToUse(tr);
       return this.m_retryAnalyzer;
     }
-    final String keyAsString = getSimpleName() + "#" + getParameterInvocationCount();
+
+    final String keyAsString = getSimpleName() + "#" + hashParameters(tr);
     return m_testMethodToRetryAnalyzer.computeIfAbsent(
         keyAsString,
         key -> {
@@ -825,6 +827,11 @@ public abstract class BaseTestMethod implements ITestNGMethod, IInvocationStatus
           CreationAttributes attributes = new CreationAttributes(tr.getTestContext(), ba, null);
           return (IRetryAnalyzer) Dispenser.newInstance(m_objectFactory).dispense(attributes);
         });
+  }
+
+  private int hashParameters(ITestResult itr) {
+    Object[] parameters = itr.getParameters();
+    return Objects.hash(parameters);
   }
 
   private static boolean isNotParameterisedTest(ITestResult tr) {

--- a/testng-core/src/main/java/org/testng/internal/invokers/ITestInvoker.java
+++ b/testng-core/src/main/java/org/testng/internal/invokers/ITestInvoker.java
@@ -1,6 +1,8 @@
 package org.testng.internal.invokers;
 
 import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
 import org.testng.IInvokedMethod;
 import org.testng.ITestContext;
 import org.testng.ITestNGMethod;
@@ -14,9 +16,9 @@ public interface ITestInvoker {
 
   class FailureContext {
 
-    int count = 0;
+    AtomicInteger count = new AtomicInteger(0);
     List<Object> instances = Lists.newArrayList();
-    boolean representsRetriedMethod = false;
+    AtomicBoolean representsRetriedMethod = new AtomicBoolean(false);
   }
 
   List<ITestResult> invokeTestMethods(

--- a/testng-core/src/main/java/org/testng/internal/invokers/MethodRunner.java
+++ b/testng-core/src/main/java/org/testng/internal/invokers/MethodRunner.java
@@ -59,13 +59,14 @@ public class MethodRunner implements IMethodRunner {
           result.addAll(tmpResults);
         } else {
           List<ITestResult> retryResults = Lists.newArrayList();
-          failure = testInvoker.retryFailed(tmArguments, retryResults, failure.count, context);
+          failure =
+              testInvoker.retryFailed(tmArguments, retryResults, failure.count.get(), context);
           result.addAll(retryResults);
         }
 
         // If we have a failure, skip all the
         // other invocationCounts
-        if (failure.count > 0
+        if (failure.count.get() > 0
             && (skipFailedInvocationCounts
                 || tmArguments.getTestMethod().skipFailedInvocations())) {
           while (invocationCount.getAndDecrement() > 0) {
@@ -122,7 +123,7 @@ public class MethodRunner implements IMethodRunner {
               context,
               skipFailedInvocationCounts,
               invocationCount.get(),
-              failure.count,
+              failure.count.get(),
               testInvoker.getNotifier());
       workers.add(w);
       // testng387: increment the param index in the bag.

--- a/testng-core/src/main/java/org/testng/internal/invokers/TestInvoker.java
+++ b/testng-core/src/main/java/org/testng/internal/invokers/TestInvoker.java
@@ -234,8 +234,8 @@ class TestInvoker extends BaseInvoker implements ITestInvoker {
       int failureCount,
       ITestContext testContext) {
     FailureContext failure = new FailureContext();
-    failure.count = failureCount;
-    failure.representsRetriedMethod = true;
+    failure.count.set(failureCount);
+    failure.representsRetriedMethod.set(true);
     do {
       failure.instances = Lists.newArrayList();
       Object[] parameterValues = arguments.getParameterValues();
@@ -533,7 +533,7 @@ class TestInvoker extends BaseInvoker implements ITestInvoker {
     } else {
       testResult.setStatus(holder.status);
       if (holder.status == ITestResult.FAILURE && !holder.handled) {
-        int count = failure.count++;
+        int count = failure.count.getAndIncrement();
         if (testMethod.isDataDriven()) {
           count = 0;
         }
@@ -580,7 +580,7 @@ class TestInvoker extends BaseInvoker implements ITestInvoker {
     long startTime = System.currentTimeMillis();
     InvokedMethod invokedMethod = new InvokedMethod(startTime, testResult);
 
-    if (!failureContext.representsRetriedMethod
+    if (!failureContext.representsRetriedMethod.get()
         && invoker.hasConfigurationFailureFor(
             arguments.getTestMethod(),
             arguments.getTestMethod().getGroups(),
@@ -827,46 +827,50 @@ class TestInvoker extends BaseInvoker implements ITestInvoker {
     return result;
   }
 
+  private static final Object LOCK = new Object();
+
   private StatusHolder considerExceptions(
       ITestNGMethod tm,
       ITestResult testResult,
       ExpectedExceptionsHolder exceptionsHolder,
       FailureContext failure) {
-    StatusHolder holder = new StatusHolder();
-    int status = testResult.getStatus();
-    holder.handled = false;
+    synchronized (LOCK) {
+      StatusHolder holder = new StatusHolder();
+      int status = testResult.getStatus();
+      holder.handled = false;
 
-    Throwable ite = testResult.getThrowable();
-    if (status == ITestResult.FAILURE && ite != null) {
+      Throwable ite = testResult.getThrowable();
+      if (status == ITestResult.FAILURE && ite != null) {
 
-      //  Invocation caused an exception, see if the method was annotated with @ExpectedException
-      if (exceptionsHolder != null) {
-        if (exceptionsHolder.isExpectedException(ite)) {
-          testResult.setStatus(ITestResult.SUCCESS);
-          status = ITestResult.SUCCESS;
-        } else {
-          if (isSkipExceptionAndSkip(ite)) {
-            status = ITestResult.SKIP;
+        //  Invocation caused an exception, see if the method was annotated with @ExpectedException
+        if (exceptionsHolder != null) {
+          if (exceptionsHolder.isExpectedException(ite)) {
+            testResult.setStatus(ITestResult.SUCCESS);
+            status = ITestResult.SUCCESS;
           } else {
-            testResult.setThrowable(exceptionsHolder.wrongException(ite));
-            status = ITestResult.FAILURE;
+            if (isSkipExceptionAndSkip(ite)) {
+              status = ITestResult.SKIP;
+            } else {
+              testResult.setThrowable(exceptionsHolder.wrongException(ite));
+              status = ITestResult.FAILURE;
+            }
           }
+        } else {
+          handleException(ite, tm, testResult, failure.count.getAndIncrement());
+          holder.handled = true;
+          status = testResult.getStatus();
         }
-      } else {
-        handleException(ite, tm, testResult, failure.count++);
-        holder.handled = true;
-        status = testResult.getStatus();
+      } else if (status != ITestResult.SKIP && exceptionsHolder != null) {
+        TestException exception = exceptionsHolder.noException(tm);
+        if (exception != null) {
+          testResult.setThrowable(exception);
+          status = ITestResult.FAILURE;
+        }
       }
-    } else if (status != ITestResult.SKIP && exceptionsHolder != null) {
-      TestException exception = exceptionsHolder.noException(tm);
-      if (exception != null) {
-        testResult.setThrowable(exception);
-        status = ITestResult.FAILURE;
-      }
+      holder.originalStatus = testResult.getStatus();
+      holder.status = status;
+      return holder;
     }
-    holder.originalStatus = testResult.getStatus();
-    holder.status = status;
-    return holder;
   }
 
   private static void updateStatusHolderAccordingToTestResult(

--- a/testng-core/src/main/java/org/testng/internal/invokers/TestMethodWithDataProviderMethodWorker.java
+++ b/testng-core/src/main/java/org/testng/internal/invokers/TestMethodWithDataProviderMethodWorker.java
@@ -74,7 +74,7 @@ public class TestMethodWithDataProviderMethodWorker implements Callable<List<ITe
     XmlSuite suite = m_testContext.getSuite().getXmlSuite();
 
     final ITestInvoker.FailureContext failure = new ITestInvoker.FailureContext();
-    failure.count = m_failureCount;
+    failure.count.set(m_failureCount);
     try {
       tmpResults.add(
           m_testInvoker.invokeTestMethod(
@@ -92,7 +92,7 @@ public class TestMethodWithDataProviderMethodWorker implements Callable<List<ITe
               suite,
               failure));
     } finally {
-      m_failureCount = failure.count;
+      m_failureCount = failure.count.get();
       if (failure.instances.isEmpty()) {
         m_testResults.addAll(tmpResults);
       } else {
@@ -100,7 +100,8 @@ public class TestMethodWithDataProviderMethodWorker implements Callable<List<ITe
           List<ITestResult> retryResults = Lists.newArrayList();
 
           m_failureCount =
-              m_testInvoker.retryFailed(
+              m_testInvoker
+                  .retryFailed(
                       new Builder()
                           .usingInstance(instance)
                           .forTestMethod(m_testMethod)
@@ -115,7 +116,8 @@ public class TestMethodWithDataProviderMethodWorker implements Callable<List<ITe
                       retryResults,
                       m_failureCount,
                       m_testContext)
-                  .count;
+                  .count
+                  .get();
           m_testResults.addAll(retryResults);
         }
       }

--- a/testng-core/src/test/java/test/dataprovider/DataProviderTest.java
+++ b/testng-core/src/test/java/test/dataprovider/DataProviderTest.java
@@ -47,6 +47,9 @@ import test.dataprovider.issue2819.TestClassSample;
 import test.dataprovider.issue2819.TestClassUsingDataProviderRetrySample;
 import test.dataprovider.issue2819.TestClassWithMultipleRetryImplSample;
 import test.dataprovider.issue2888.SkipDataProviderSample;
+import test.dataprovider.issue2934.TestCaseSample;
+import test.dataprovider.issue2934.TestCaseSample.CoreListener;
+import test.dataprovider.issue2934.TestCaseSample.ToggleDataProvider;
 
 public class DataProviderTest extends SimpleBaseTest {
 
@@ -591,5 +594,79 @@ public class DataProviderTest extends SimpleBaseTest {
     testNG.addListener(listener);
     testNG.run();
     assertThat(listener.getParameters()).containsExactlyElementsOf(Arrays.asList(1, 2, 3, 4, 5));
+  }
+
+  @Test(description = "GITHUB-2934")
+  public void ensureParallelDataProviderWithRetryAnalyserWorks() {
+    runTest(true);
+  }
+
+  @Test(description = "GITHUB-2934")
+  public void ensureSequentialDataProviderWithRetryAnalyserWorks() {
+    runTest(false);
+  }
+
+  private static void runTest(boolean isParallel) {
+    TestNG testng = create(TestCaseSample.class);
+    CoreListener listener = new CoreListener();
+    ToggleDataProvider transformer = new ToggleDataProvider(isParallel);
+    testng.addListener(listener);
+    testng.addListener(transformer);
+    testng.setVerbose(2);
+    testng.run();
+    SoftAssertions softly = new SoftAssertions();
+
+    softly
+        .assertThat(listener.totalTests())
+        .withFailMessage("We should have had 9 test results in total.")
+        .isEqualTo(9);
+
+    // Assert passed tests
+    softly
+        .assertThat(listener.getPassedTests())
+        .withFailMessage("We should have had ONLY 1 passed test.")
+        .hasSize(1);
+    listener
+        .getPassedTests()
+        .forEach(
+            each ->
+                softly
+                    .assertThat(each.wasRetried())
+                    .withFailMessage(
+                        "Passed test " + stringify(each) + " should NOT have been retried")
+                    .isFalse());
+
+    // Assert failed tests
+    assertThat(listener.getFailedTests())
+        .withFailMessage("We should have had 2 failed tests.")
+        .hasSize(2);
+    listener
+        .getFailedTests()
+        .forEach(
+            each ->
+                softly
+                    .assertThat(each.wasRetried())
+                    .withFailMessage(
+                        "Failed test " + stringify(each) + " should NOT have been retried")
+                    .isFalse());
+
+    // Assert skipped tests
+    assertThat(listener.getSkippedTests())
+        .withFailMessage("We should have had 6 skipped tests due to retries.")
+        .hasSize(6);
+    listener
+        .getSkippedTests()
+        .forEach(
+            each ->
+                softly
+                    .assertThat(each.wasRetried())
+                    .withFailMessage(
+                        "Skipped test " + stringify(each) + " should have been retried")
+                    .isTrue());
+    softly.assertAll();
+  }
+
+  private static String stringify(ITestResult itr) {
+    return "(" + Arrays.toString(itr.getParameters()) + ")";
   }
 }

--- a/testng-core/src/test/java/test/dataprovider/issue2934/TestCaseSample.java
+++ b/testng-core/src/test/java/test/dataprovider/issue2934/TestCaseSample.java
@@ -1,0 +1,82 @@
+package test.dataprovider.issue2934;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.lang.reflect.Method;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.testng.IAnnotationTransformer;
+import org.testng.IRetryAnalyzer;
+import org.testng.ITestContext;
+import org.testng.ITestListener;
+import org.testng.ITestResult;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.IDataProviderAnnotation;
+import org.testng.annotations.Test;
+
+public class TestCaseSample {
+
+  @Test(dataProvider = "testData", retryAnalyzer = RetryAnalyzer.class)
+  public void API_dummyTest(int number) {
+    assertThat(number % 2).isEqualTo(0);
+  }
+
+  @DataProvider(name = "testData")
+  public Object[][] getData() {
+    return new Object[][] {{1}, {2}, {3}};
+  }
+
+  public static class ToggleDataProvider implements IAnnotationTransformer {
+
+    private final boolean isParallel;
+
+    public ToggleDataProvider(boolean isParallel) {
+      this.isParallel = isParallel;
+    }
+
+    @Override
+    public void transform(IDataProviderAnnotation annotation, Method method) {
+      if (isParallel) {
+        annotation.setParallel(true);
+      }
+    }
+  }
+
+  public static class RetryAnalyzer implements IRetryAnalyzer {
+
+    private final AtomicInteger counter = new AtomicInteger(0);
+
+    @Override
+    public boolean retry(ITestResult iTestResult) {
+      return counter.getAndIncrement() < 3;
+    }
+  }
+
+  public static class CoreListener implements ITestListener {
+
+    private ITestContext context;
+
+    @Override
+    public void onFinish(ITestContext context) {
+      this.context = context;
+    }
+
+    public int totalTests() {
+      return context.getPassedTests().size()
+          + context.getFailedTests().size()
+          + context.getSkippedTests().size();
+    }
+
+    public Set<ITestResult> getPassedTests() {
+      return context.getPassedTests().getAllResults();
+    }
+
+    public Set<ITestResult> getFailedTests() {
+      return context.getFailedTests().getAllResults();
+    }
+
+    public Set<ITestResult> getSkippedTests() {
+      return context.getSkippedTests().getAllResults();
+    }
+  }
+}

--- a/testng-core/src/test/java/test/retryAnalyzer/RetryAnalyzerTest.java
+++ b/testng-core/src/test/java/test/retryAnalyzer/RetryAnalyzerTest.java
@@ -49,7 +49,10 @@ public class RetryAnalyzerTest extends SimpleBaseTest {
     Map<Integer, Long> collected =
         HashCodeAwareRetryAnalyzer.hashCodes.stream()
             .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
-    assertThat(collected.keySet()).hasSize(1);
+    assertThat(collected.keySet())
+        .withFailMessage(
+            "There should have been 2 RetryAnalyser instances since the data driven test has 2 sets of data")
+        .hasSize(2);
   }
 
   @Test

